pallets/aggregate/src/lib.rs

```diff
@@ -0,0 +1,608 @@
+        }
+    }
+
+    fn available_aggregation<T: Config>(domain: &mut Domain<T>, aggregation: Aggregation<T>) {
+        Pallet::<T>::deposit_event(Event::<T>::ReadyToAggregate {
+            domain_id: domain.id,
+            aggregation_id: aggregation.id,
+        });
+        domain
+            .should_publish
+            .try_insert(aggregation.id, aggregation)
+            .expect("Should not publish aggregation if it's not possible: qed");
+        // If is full send an alert event
+        if domain.should_publish.len() >= domain.publish_queue_size as usize {
+            Pallet::<T>::deposit_event(Event::<T>::DomainFull {
+                domain_id: domain.id,
+            });
+        }
+    }
+
+    // Errors inform users that something went wrong.
+    #[pallet::error]
+    pub enum Error<T> {
+        /// This domain id is unknown.
+        UnknownDomainId,
+        /// This aggregation cannot be published or it's already published.
+        InvalidAggregationId,
+        /// The domain params are invalid.
+        InvalidDomainParams,
+    }
+
+    #[derive(Debug, Clone, PartialEq, Encode, Decode, TypeInfo, MaxEncodedLen)]
+    pub enum CannotAggregateCause {
+        NoAccount,
+        NoDomain,
+        DomainNotRegistered { domain_id: u32 },
+        DomainStorageFull { domain_id: u32 },
+        InsufficientFound,
+    }
+
+    #[pallet::event]
+    #[pallet::generate_deposit(pub(super) fn deposit_event)]
+    pub enum Event<T: Config> {
+        NewDomain {
+            id: u32,
+        },
+        NewProof {
+            statement: H256,
+            domain_id: u32,
+            aggregation_id: u64,
+        },
+        ReadyToAggregate {
+            domain_id: u32,
+            aggregation_id: u64,
+        },
+        NewAggregationReceipt {
+            domain_id: u32,
+            aggregation_id: u64,
+            receipt: H256,
+        },
+        AggregationRemoved {
+            domain_id: u32,
+            aggregation_id: u64,
+        },
+        CannotAggregate {
+            statement: H256,
+            cause: CannotAggregateCause,
+        },
+        DomainFull {
+            domain_id: u32,
+        },
+    }
+
+    /// Shortcut to get the Aggregation type from config.
+    pub type Aggregation<T> =
+        crate::data::AggregationEntry<AccountOf<T>, BalanceOf<T>, <T as Config>::AggregationSize>;
+
+    /// Shortcut to get the Domain type from config.
+    pub(crate) type Domain<T> = crate::data::DomainEntry<
+        AccountOf<T>,
+        BalanceOf<T>,
+        <T as Config>::AggregationSize,
+        <T as Config>::MaxPendingPublishQueueSize,
+        TicketOf<T>,
+    >;
+
+    /// Compute and reserve the currency for further publication
+    fn reserve_currency_for_publication<T: Config>(
+        domain: &mut Domain<T>,
+        account: &AccountOf<T>,
+    ) -> Result<BalanceOf<T>, DispatchError> {
+        let estimated = estimate_publish_attestation_fee::<T>(domain.max_aggregation_size);
```

Comment:
Is there a particular reason we don't allow publishing incomplete aggregations ?

Reply:
Yes, because the proof owner just hold the balance for their contribute on the base that this is complete. If we try to publish it before there could be not enough founds to cover the price.

Reply:
Ummm can't we assume the tip is covering also for that ? Pulisher can decided when it's profitable or not. If he envisions publishing non-complete aggregations he can ask for more tip